UC-206, UC-82 Graceful handling of unreachable Facebook + Remove FB lib load on every page

armonr opened the pull request:
This PR handles the case when Facebook is not available (blocked, or down) by:
- hiding FB login / signup controls throughout all login/signup page and global navigation
- Throwing an error to user on an attempt to navigate to their FB user preference tab.

Also it is removing load of Facebook on every page load. It is being done asynchronously.

Also loading Facebook is no longer using the `getResources` function because "cross site script loading causes `$.getScript` to not execute fail functions." Thanks @lizlux for this fix. More info: #5939

https://wikia-inc.atlassian.net/browse/UC-206
https://wikia-inc.atlassian.net/browse/UC-82

Commit:
UC-206 Clean up getResources script

Commit:
Merge branch 'uconn-4' into UC-206

Commit:
UC-206 Add messages for FB fail

Commit:
UC-206 Styling changes to allow hiding FB buttons everywhere

Commit:
UC-206 Add FB element classes + Add messages to JS

Commit:
UC-206 Syntax fix in getResources

Commit:
UC-206 UC-82 Stop loading FB on every page, and
Callback should happen by caller of loadFacebookAPI()

Commit:
UC-206 Add classes to user pref tabs and custom events

Commit:
UC-206 All callers of .loadFacebookAPI() need to make callbacks on done()
Also implement Error on FB user pref tab - to be completed

lizlux committed:
UC-206 fixing facebook sdk loading with proper error case handling

lizlux committed:
UC-206 whitespace

armonr committed:
Merge pull request #5939 from Wikia/UC-206-liz

UC-206 Fail state will trigger while loading fb

Commit:
UC-206 Fix FB lib fail case after FB lib load fix by @lizlux:
https://github.com/Wikia/app/pull/5939

Commit:
UC-206 Disable FB login for Mobile web if FB unavailable

lizlux commented on extensions/wikia/CreateNewWiki/js/CreateNewWiki.js:
```diff
@@ -88,10 +88,17 @@
+							$('.fb-loaded').removeClass('hidden');
+						})
+						.fail(function () {
+							// you can do something here if you want
```
Hehe, I was expecting this line to be replaced with code, but since it's not, let's remove the `.fail()` call.

lizlux commented on extensions/wikia/FacebookClient/scripts/preferences.js:
```diff
@@ -24,7 +24,17 @@
+			$.loadFacebookAPI()
+				.done(function () {
+					$('.fb-loaded').removeClass('hidden');
+
```
I would get rid of this space but that's a personal preference.

lizlux commented on extensions/wikia/UserLogin/templates/UserLoginSpecial_Providers.php:
```diff
@@ -1,4 +1,4 @@
-	<div class="sso-login">
+	<div class="sso-login fb-loaded hidden">
```
This class seems like a state marker, and by default (i.e. when this html is loaded), facebook is not yet loaded. Is there any reason we can't use `.sso-login`? That element might even be already cached in some cases.

armonr replied:
You have a good point about it being misleading. I did this to separate styling class (i.e. `.sso-login`) from the functional class `fb-loaded`. Rename?

lizlux replied:
Yeah, separating classes that are for function vs style is a good idea and we don't do much of that at wikia. Renaming would be fine. How about `fb-pending`? Then you can remove that class and the hidden class at the same time.

lizlux commented on resources/mediawiki.special/mediawiki.special.preferences.js:
```diff
@@ -70,6 +85,10 @@ $legends.each( function( i, legend ) {
 	$preftoc.append( $li );
+
+	/** Wikia Change start - Add a custom event per tab - @see UC-206 */
+	$(document).trigger($tabName + '-complete');
```
I think this can be simplified a bit. Currently, this is firing an event per tab, which is unnecessary and could slow things down if there were a bunch of tabs. `$.each()` is synchronous, so you could just fire one event after it's done looping. 

What would simplify this more, though, would be to fire an event when the tab is actually clicked. That way, you could reduce the event binding in preferences.js to just one click event, instead of a load event and a click event.

armonr replied:
So simpler alternative would be to bind event to clicks for each tab. That will eliminate the second level of events. Let me know if there is a better way.

lizlux commented on extensions/wikia/FacebookClient/scripts/preferences.js:
```diff
@@ -134,6 +144,42 @@
+
+					// Disable everything within the tab
+					$tabContentContainer
+						.find('input, button')
```
This seems unnecessary. There's no form elements on this page and there likely will never be.

lizlux commented on extensions/wikia/FacebookClient/scripts/preferences.js:
```diff
@@ -134,6 +144,42 @@
+					function createModal(uiModal) {
+						var modalConfig = {
+							vars: {
+								id: 'FbErrorModal',
```
The naming convention is actually lowerCamelCase for ids https://github.com/Wikia/guidelines/blob/master/SCSS/CodingConventions.md#class-and-id-naming-conventions

lizlux commented on resources/wikia/libraries/jquery/getResources/jquery.wikia.getResources.js:
```diff
@@ -1,101 +1,110 @@
+	 */
+	var getResources = (function () {
+		var rAssetManagerGroup = new RegExp(window.wgAssetsManagerQuery
+					// Escape any special regex characters in the URL
```
these should actually only be indented by 1 tab, but IDE's don't seem to comprehend that.

bkoval commented on extensions/wikia/FacebookClient/scripts/preferences.js:
```diff
@@ -24,7 +24,17 @@
+					bindEvents();
+				})
+				.fail(function () {
+					facebookError();
```
Could this be .fail(facebookError); ?

armonr replied:
Correct, will fix :)

bkoval commented on extensions/wikia/UserLogin/css/UserLogin.scss:
```diff
@@ -75,3 +75,7 @@ $input-width: 230px;
 	}
 }
+
+.hidden {
```
Hey, are you considering to put this class into a shared file, according to your question in https://wikia-inc.atlassian.net/wiki/questions/19791892/global-css-rules ? That would be great and we could all benefit from that I guess :)

armonr replied:
@bkoval I would really like to do that :) However from my understanding, after talking with @lizlux there may be lots of cases where this rule gets overwritten with local rule with `!important` and it may contain slightly different rules.

If all @wikia-frontenders agree on making this happen, I would be super happy to take that on!

lizlux replied:
I didn't say this rule gets overridden with `!important`, I said if we want to set a global `.hidden` rule, we'd want it to include `!important` so it would override more specific rules like `.my-module .my-div { display: none; }

I think we should create a separate ticket to do this work and ideally replace the per-extension instances of `.hidden`. We've done a fairly good job of sticking to the `.hidden` class name so it shouldn't be too hard to update.

lizlux replied:
https://wikia-inc.atlassian.net/browse/UC-224

armonr replied:
:+1:

bkoval replied:
Yup, it's a great use case of !important and it should be there to override any stronger selectors / inline styles. @armonr :+1:

bkoval commented on extensions/wikia/UserLogin/css/UserLoginDropdown.globalNavigation.scss:
```diff
@@ -54,7 +54,7 @@
 
 	.WikiaForm {
-		padding: 0 $large-content-padding;
+		padding: 0 $large-content-padding $large-content-padding $large-content-padding;
```
@armonr @lizlux What do you think, would:
padding: {
	bottom: $large-content-padding;
	left: $large-content-padding;
	right: $large-content-padding;
	top: 0;
}
make things slightly clearer or it's not the best idea? :)

armonr replied:
This is slightly more compacted form, but your suggestion is much more clear for sure. There are other instances that can be changed in this file. What's your take @lizlux ?

lizlux replied:
Our guidelines specify it should be: 

`padding: 0 $large-content-padding $large-content-padding`

The 3rd `$large-content-padding` is redundant b/c css will take the already specified "right" value and apply it to the "left" value.

Commit:
UC-206 Trigger onclick events for user pref tabs, and
address other feedback from @lizlux

lizlux commented on extensions/wikia/FacebookClient/scripts/preferences.js:
```diff
@@ -24,7 +24,14 @@
-			$.loadFacebookAPI(bindEvents);
+			$.loadFacebookAPI()
+				.done(function () {
+					$('.fb-loaded').removeClass('hidden');
```
In this case, the `$connectLink` is already cached, so we should be querying the DOM for it

lizlux commented on extensions/wikia/UserLogin/js/UserLoginFacebook.js:
```diff
@@ -58,7 +58,10 @@
-				$.loadFacebookAPI();
+				$.loadFacebookAPI()
+					.done(function () {
+						$('.fb-loaded').removeClass('hidden');
```
Same here, we're querying the DOM in line 79, so we should duplicate that query.

Commit:
UC-206 Remove redundant left padding per @lizlux

Commit:
U -nC-206 Remove class fb-loaded in favor of existing sso-login
This is because the class is ultimately used for styling which will be redundant.
CC @lizlux

lizlux commented on extensions/wikia/UserLogin/css/UserLoginDropdown.globalNavigation.scss:
```diff
@@ -54,7 +54,7 @@
 
 	.WikiaForm {
-		padding: 0 $large-content-padding;
+		padding: 0 $large-content-padding $large-content-padding;
```
I just checked this out and it looks like there's some extra padding in the drop down now that wasn't there before
![screen shot 2015-01-14 at 2 26 50 pm](https://cloud.githubusercontent.com/assets/201367/5749069/24bd0e3e-9bfa-11e4-954c-d2f3b2e96fe7.png)
![screen shot 2015-01-14 at 2 32 26 pm](https://cloud.githubusercontent.com/assets/201367/5749075/304930ac-9bfa-11e4-8a48-9c25cee8711c.png)

lizlux commented on resources/mediawiki.special/mediawiki.special.preferences.js:
```diff
@@ -19,6 +19,18 @@ var $legends = $fieldsets.children( 'legend' )
 	.addClass( 'mainLegend' );
 
 	/**
+	 * Wikia Function - @see UC-206
```
It looks like the indentation is off for these lines https://github.com/Wikia/app/blob/UC-206/resources/mediawiki.special/mediawiki.special.preferences.js#L21-L47

armonr replied:
Right, because the whole file is wrapped in a function but not indented. I will indent these back!

Commit:
UC-206 Fix FB padding in navigation

Commit:
Fix indentation

lizlux commented:
LGTM!
